### PR TITLE
Define and Implement C++ API for negative sampling

### LINKED ISSUE #4497: Implement negative sampling in cugraph-core
Negative sampling is one of the features we need for GNN support.

PyG implementation is here as a reference:  https://github.com/pyg-team/pytorch_geometric/blob/845c5cffc086c0a37dd3255d5f9219d982439efe/torch_geometric/sampler/base.py#L574

### PR DESCRIPTION
**Pull request by ChuckHastings**
Defines and implements the PLC/C/C++ APIs for negative sampling.

Closes #4497

### CODE REVIEW

File: cpp/tests/sampling/mg_negative_sampling.cu [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+  current_test->run_current_test(
+    rng_state, Negative_Sampling_Usecase{2, true, false, false, false, true, true});
+  current_test->run_current_test(
+    rng_state, Negative_Sampling_Usecase{2, false, true, false, false, true, true});
+  current_test->run_current_test(
+    rng_state, Negative_Sampling_Usecase{2, true, true, false, false, true, true});
+  current_test->run_current_test(
+    rng_state, Negative_Sampling_Usecase{2, false, false, true, false, true, true});
+  current_test->run_current_test(
+    rng_state, Negative_Sampling_Usecase{2, true, false, true, false, true, true});
```
Comment (ChuckHastings) [inline cpp/tests/sampling/mg_negative_sampling.cu:292]: @seunghwak - When I run this test with np = 2 I'm getting the error in shuffling (np = 1 works fine).  Haven't tried disabling just this test to see if any tests after this fail.

**Review by alexbarghi-nv — APPROVED** [review]
👍

File: cpp/tests/sampling/mg_negative_sampling.cu (hunk 1 above)
Comment (ChuckHastings) [reply cpp/tests/sampling/mg_negative_sampling.cu:287]: Fixed

**Review by jnke2016 — APPROVED** [review]

File: cpp/src/sampling/negative_sampling_impl.cuh [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    if constexpr (multi_gpu) {
+      weight_t dst_bias_sum{0};
+
+      if (dst_bias) {
+        // Compute the dst_bias sum for this partition and normalize cached values
+        dst_bias_sum = thrust::reduce(
+          handle.get_thrust_policy(),
+          dst_bias_cache_->view().value_first(),
+          dst_bias_cache_->view().value_first() + graph_view.local_edge_partition_dst_range_size(),
+          weight_t{0});
```
Comment (seunghwak) [inline cpp/src/sampling/negative_sampling_impl.cuh:122]: This won't work.

edge_src|dst_property_t stores edge src|dst property values in either linear array or (key, value) pairs.

This works only if property values are store in an linear array, but with this approach, memory footprint scales as a function of O(V/sqrt(P)), so this won't scale in large scale for graphs with a relative low average vertex degree. Using key, value pairs ensure that memory footprint stays as min(O(V/sqrt(P)), O(E/P)).

And I am uncomfortable in exploiting this implementation details outside the primitives. And if we are only visiting the edges that exist in the input graph, we can store in (key, value) pairs, but if we need to cover the entire src/dst range, edge_property_t is not a right data structure.

One robust (but wastes some communication) approach is this.

We locally sum src biases and dst biases in each rank, then call host_scalar_allgather.

We first generate edge sources in two steps.

1) In each rank, find how many sources to generate in each V/P segment (host_scalar_allgather output of local src bias sum). Then call host_scalar_all_to_all (host_scalar_multicast). Based on this, each rank generates sources for its local vertex partition range. Then, send the source values back.
2) Do the same for destinations.
3) Shuffle edges.

A key draw back is that we need to send sources & destinations back and shuffle edges as well. More communication than necessary but no memory footprint issue and minimally exploits any cugraph implementation details.

A more efficient approach is to create a P by sqrt(P) (or minor_comm_size to be exact) array (this requires understanding how the adjacency matrix is partitoined). By just collecting local bias sum values, we can compute how many edges should be generated in each box. We can send these nuumbers to the rank owning the boxes.

When we generate sources, we can generate sources in minor_com_size iterations; in each iteration we hold source biases only for the V/P vertices. When we generate destination vertices. We first just compute how many to generate per each V/P range in each box. Sum it, send to the owning GPUs (or in loops, we can bring V/P dst bias values based on which incur less communication), and get the destination vertices back. This will incur less communication but more involved.

I am inclined to implement the first approach, and if that turns out to be a performance bottleneck, we may consider implementing the second approach may be in the primitive space.... Let me know what you think.

Comment (ChuckHastings) [reply cpp/src/sampling/negative_sampling_impl.cuh:122]: I'll revisit this.

My first attempt at refactoring did something akin to your first suggestion.  I ran into some problems in debugging and then the current approach occurred to me.  My second attempt was something akin to your second approach, but then I ran into the issue that the source edge partitions are separate.  This was my third attempt and works [as long as the internal implementation is an array and not a k-v implementation... I guess I didn't have a test that reached that point].

I agree that doing this properly and efficiently might need some work in primitive space.

File: cpp/include/cugraph/detail/utility_wrappers.hpp [hunk 3]
```diff
@@ -50,6 +50,29 @@ void uniform_random_fill(rmm::cuda_stream_view const& stream_view,
...
+ * @param[in]   rng_state    The RngState instance holding pseudo-random number generator state.
+ * @param[out]  output       The random values
+ * @param[in]   biases       The biased values
+ *
+ */
+template <typename value_t, typename bias_t>
+void biased_random_fill(raft::handle_t const& handle,
+                        raft::random::RngState& rng_state,
+                        raft::device_span<value_t> output,
+                        raft::device_span<bias_t const> biases);
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/include/cugraph/detail/utility_wrappers.hpp:74]: I may delete this function.  I implemented a different way.  After I've refactored the implementation I'll revisit whether we need this function or not.  If we keep it I'll make that change.

File: cpp/include/cugraph/sampling_functions.hpp [hunk 4]
```diff
@@ -743,4 +743,60 @@ lookup_endpoints_from_edge_ids_and_types(
...
+ *
+ * This function generates negative samples for graph.
+ *
+ * Negative sampling is done by generating a random graph according to the specified
+ * parameters and optionally removing the false negatives.
+ *
+ * Sampling occurs by creating a list of source vertex ids from biased samping
+ * of the source vertex space, and destination vertex ids from biased sampling of the
+ * destination vertex space, and using this as the putative list of edges.  We
+ * then can optionally remove duplicates and remove false negatives to generate
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/include/cugraph/sampling_functions.hpp:757]: Missed the change in the documentation... I'll search the documentation and push a fix once I resolve the testing issue on one of the build configurations.

File: cpp/src/sampling/negative_sampling_impl.cuh (hunk 2 above)
Comment (ChuckHastings) [reply cpp/src/sampling/negative_sampling_impl.cuh]: Refactored in latest push to use option 1.  Note that's only required for biased... if either src or dst is uniform that can be done directly.

I suspect the original problem I had with that approach was actually the raft bug that I later worked around.  Code is much simpler this way, and I think the performance should be fine - at least for now.

File: cpp/src/sampling/negative_sampling_impl.cuh [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+      thrust::adjacent_difference(
+        handle.get_thrust_policy(), gpu_counts.begin(), gpu_counts.end(), gpu_counts.begin());
+
+      // all_gpu_counts[i][j] will be how many vertices need to be generated on GPU j to be sent to
+      // GPU i
+      auto all_gpu_counts = cugraph::device_allgatherv(
+        handle,
+        handle.get_comms(),
+        raft::device_span<size_t const>{gpu_counts.data(), gpu_counts.size()});
```
Comment (seunghwak) [inline cpp/src/sampling/negative_sampling_impl.cuh:166]: You need to just collect sample counts for comm_size GPUs (no need to collect for the entire comm_size * comm_size matrix). Use device_multicast_sendrecv to send to comm_size GPUs and receive from comm_size GPUs.

Comment (seunghwak) [reply cpp/src/sampling/negative_sampling_impl.cuh:166]: Then, the code below will also become simpler.

**Review by jnke2016 — APPROVED** [review]

**Comment by jnke2016** [comment]
I re-reviewed the refactored PR and approve it

**Review by ChuckHastings — COMMENTED** [review]
Latest push addresses the latest batch of requests.  Please review again.

**Review by seunghwak — APPROVED** [review]
LGTM besides minor cosmetic issues and the need to test with edge masked graphs (to make sure that every algorithm works properly with edge masked graphs).

File: cpp/tests/sampling/mg_negative_sampling.cpp [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+#include <gtest/gtest.h>
+
+struct Negative_Sampling_Usecase {
+  float sample_multiplier{2};
+  bool use_src_bias{false};
+  bool use_dst_bias{false};
+  bool remove_duplicates{false};
+  bool remove_existing_edges{false};
+  bool exact_number_of_samples{false};
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/tests/sampling/mg_negative_sampling.cpp:33]: What would behavior be in an edge masked graph?  That if an edge is masked out it would be allowed to be in the negative sampled edges if `remove_existing_edges` is set to true?  I can add that tests for that.

[the remark this answers is not in the document]
Comment (seunghwak) [reply cpp/tests/sampling/mg_negative_sampling.cpp:33]: Yes. If you call has_edge() on masked out edges, you will get false. So, those edges can be included in the negative sampling output

File: cpp/include/cugraph/sampling_functions.hpp [hunk 7]
```diff
@@ -743,4 +743,61 @@ lookup_endpoints_from_edge_ids_and_types(
...
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> negative_sampling(
+  raft::handle_t const& handle,
+  raft::random::RngState& rng_state,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  size_t num_samples,
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/include/cugraph/sampling_functions.hpp:795]: Moved in next push

File: cpp/src/sampling/negative_sampling_impl.cuh [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+                    normalized_biases->begin(),
+                    divider_t<weight_t>{sum});
+
+  thrust::inclusive_scan(handle.get_thrust_policy(),
+                         normalized_biases->begin(),
+                         normalized_biases->end(),
+                         normalized_biases->begin());
+
+  if constexpr (multi_gpu) {
+    // rmm::device_scalar<weight_t> d_sum((sum / aggregate_sum), handle.get_stream());
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/src/sampling/negative_sampling_impl.cuh:77]: Deleted in next push

File: cpp/tests/sampling/mg_negative_sampling.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/tests/sampling/mg_negative_sampling.cpp:33]: Done in next push

File: cpp/tests/sampling/negative_sampling.cpp [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+#include <gtest/gtest.h>
+
+struct Negative_Sampling_Usecase {
+  float sample_multiplier{2};
+  bool use_src_bias{false};
+  bool use_dst_bias{false};
+  bool remove_duplicates{false};
+  bool remove_existing_edges{false};
+  bool exact_number_of_samples{false};
+  bool check_correctness{true};
```
[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/tests/sampling/negative_sampling.cpp:35]: Done in next push
